app/validators/date_validator.rb

```diff
@@ -1,12 +1,20 @@
 class DateValidator < ActiveModel::EachValidator
   def validate_each(record, attribute, value)
-    record.errors.add(attribute, error_message) unless parsable_date?(value)
+    if incorrect_format?(value)
+      record.errors.add(attribute, error_message)
+    elsif !parsable_date?(value)
+      record.errors.add(attribute, invalid_date_error_message)
+    end
   end
 
 private
 
+  def incorrect_format?(date_string)
+    !(iso8601_regex === date_string)
+  end
+
   def parsable_date?(date_string)
-    date_string =~ iso8601_regex && Date.parse(date_string)
+    iso8601_regex === date_string && Date.parse(date_string)
```

Comment:
This is repeating the implementation of `incorrect_format?`, not sure that is necessary.